File: examples/work-management-django/billing_seam/seam.py
"""The application-local billing seam (BC-US-151, INV-030/031).

Every plan/entitlement question the product asks goes through this module
and nothing else. In standalone mode the answers come from the local
fixtures below; the future Billing Core integration replaces the provider
behind `get_provider()` with a GraphQL-backed one WITHOUT touching any
caller — and the standalone provider (and its tests) remain forever.

Billing model (SPEC): tiered active-member pricing plus metered automation
runs with a monthly included quantity and graduated overage tiers.
Amounts are integer minor units (DKK øre) — never floats.
"""

from dataclasses import dataclass
from datetime import date


# --- Local showcase fixtures -------------------------------------------------

SEAT_TIERS = [
    # (max active members inclusive, price per member per month, tier name)
    (5, 4_900, "Studio"),
    (20, 3_900, "Agency"),
    (None, 2_900, "Scale"),
]

INCLUDED_AUTOMATION_RUNS = 200

OVERAGE_TIERS = [
    # (runs in this graduated band, price per run) — None = unbounded tail
    (300, 50),
    (500, 35),
    (None, 20),
]

CURRENCY = "DKK"


@dataclass
class UsageSummary:
    active_members: int
    seat_tier_name: str
    seat_price_minor: int
    seat_total_minor: int
    automation_runs: int
    included_runs: int
    overage_runs: int
    overage_total_minor: int
    currency: str = CURRENCY

    @property
    def total_minor(self):
        return self.seat_total_minor + self.overage_total_minor


class LocalFixtureProvider:
    """Standalone answers from the fixtures above (no network, ever)."""

    def seat_tier(self, active_members):
        for cap, price, name in SEAT_TIERS:
            if cap is None or active_members <= cap:
                return name, price
        raise AssertionError("unreachable: last tier is unbounded")

    def overage_cost_minor(self, overage_runs):
        """Graduated: each band prices only the runs inside it."""
        remaining, total = overage_runs, 0
        for band, price in OVERAGE_TIERS:
            if remaining <= 0:
                break
            in_band = remaining if band is None else min(remaining, band)
            total += in_band * price
            remaining -= in_band
        return total

    def summarize(self, organization, period_start=None):
        from automation.models import AutomationRun
        from accounts.models import Membership

        period_start = period_start or date.today().replace(day=1)
        active = organization.memberships.filter(status=Membership.ACTIVE).count()
        tier_name, seat_price = self.seat_tier(active)

        runs = AutomationRun.objects.filter(
            organization=organization,
            status=AutomationRun.SUCCEEDED,
            ran_at__date__gte=period_start,
        ).count()
        overage = max(runs - INCLUDED_AUTOMATION_RUNS, 0)

        return UsageSummary(
            active_members=active,
            seat_tier_name=tier_name,
            seat_price_minor=seat_price,
            seat_total_minor=active * seat_price,
            automation_runs=runs,
            included_runs=INCLUDED_AUTOMATION_RUNS,
            overage_runs=overage,
            overage_total_minor=self.overage_cost_minor(overage),
        )


_provider = LocalFixtureProvider()


def get_provider():
    """The single seam callers use; integrated mode swaps the provider
    for the Billing Core-backed one (BC-US-151 final milestone) without
    touching any caller."""
    from billing_core import integration

    if integration.enabled():
        from billing_core.provider import BillingCoreProvider

        return BillingCoreProvider()
    return _provider


def format_minor(amount_minor, currency=CURRENCY):
    major, minor = divmod(amount_minor, 100)
    return f"{major:,}.{minor:02d} {currency}"

File: examples/employee-directory-laravel/app/Billing/Seam.php
<?php

namespace App\Billing;

use App\Models\Organization;

/**
 * The application-local billing seam (BC-US-152, INV-030/031).
 *
 * Every plan/entitlement question goes through Seam::provider() and
 * nothing else; standalone answers come from the fixtures below, and the
 * future Billing Core integration swaps the provider without touching a
 * single caller. Amounts are integer oere — never floats.
 *
 * Billing model: ANNUAL PREPAID per active employee with a minimum seat
 * commitment; optional onboarding-workflows and advanced-directory
 * add-ons; mid-term employee growth prorates prospectively (the summary
 * exposes the prospective-proration amount for seats added mid-term).
 */
final class Seam
{
    public const PER_EMPLOYEE_YEAR_ORE = 59_900;      // per active employee per year
    public const ONBOARDING_ADDON_YEAR_ORE = 99_000;  // flat per year
    public const ADVANCED_ADDON_YEAR_ORE = 149_000;   // flat per year
    public const CURRENCY = 'DKK';

    private static ?object $provider = null;

    public static function provider(): object
    {
        if (\App\BillingCore\Integration::enabled()) {
            return self::$provider ??= new \App\BillingCore\Provider();
        }

        return self::$provider ??= new LocalFixtureProvider();
    }

    public static function swapProvider(object $provider): void
    {
        self::$provider = $provider;
    }

    public static function formatOre(int $ore): string
    {
        $major = intdiv($ore, 100);
        $minor = $ore % 100;

        return number_format($major).'.'.str_pad((string) $minor, 2, '0', STR_PAD_LEFT).' '.self::CURRENCY;
    }
}

final class LocalFixtureProvider
{
    public function summarize(Organization $organization): array
    {
        $active = $organization->activeEmployees();
        $billable = max($active, $organization->minimum_seats);
        $seats = $billable * Seam::PER_EMPLOYEE_YEAR_ORE;
        $addons = ($organization->onboarding_addon ? Seam::ONBOARDING_ADDON_YEAR_ORE : 0)
            + ($organization->advanced_directory_addon ? Seam::ADVANCED_ADDON_YEAR_ORE : 0);

        // Mid-term growth prorates prospectively: seats above the minimum
        // added today are charged for the remaining fraction of the year.
        $dayOfYear = (int) now()->format('z');
        $remainingFraction = (365 - $dayOfYear) / 365;
        $growth = max($active - $organization->minimum_seats, 0);
        $prospective = (int) floor($growth * Seam::PER_EMPLOYEE_YEAR_ORE * $remainingFraction);

        return [
            'active_employees' => $active,
            'minimum_seats' => $organization->minimum_seats,
            'billable_seats' => $billable,
            'seat_total_ore' => $seats,
            'addon_total_ore' => $addons,
            'prospective_growth_ore' => $prospective,
            'year_total_ore' => $seats + $addons,
            'currency' => Seam::CURRENCY,
        ];
    }
}
